connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java

```diff
@@ -267,6 +267,18 @@ public String memberId() {
         return JoinGroupRequest.UNKNOWN_MEMBER_ID;
     }
 
+    @Override
+    protected void handlePollTimeoutExpiry() {
```

Comment:
Since we (as maintainers) don't have good insight into what commonly causes the herder tick thread to block and the poll timeout to fire, we recently added https://issues.apache.org/jira/browse/KAFKA-15563 to help users debug it themselves.

It would be nice to integrate with this system to have the heartbeat thread report what the herder tick thread was blocked on at the time that the poll timeout happened, as this would report stalling that isn't caused by REST requests.

The integration is tricky though, because the WorkerCoordinator is (and should be) unaware of the DistributedHerder. And currently I think the WorkerCoordinator hides these internal disconnects and reconnects inside of the poll method. Perhaps we can extend the WorkerRebalanceListener or have a new error listener to allow the herder to be informed about these errors.

Reply:
Thanks Greg, I think that makes sense. I have extended the `WorkerRebalanceListener` to add another hook upon poll timeout expiry which the coordinator invokes. This way, it is able to access the tick thread's stage which was being executed at that point of time. I am just thinking, could there be race conditions where what we get in the tick thread stage might not always reflect the point where tick thread is blocked (like it moved on by the time we invoke this or becomes null), but I guess it should be ok because this can be treated as best effort? WDYT?

Reply:
Yeah the tick thread stage is a best effort and can be misleading. However, if something has blocked long enough to cause a poll timeout, its likely enough to continue for the additional time it requires for the heartbeat thread to notice.

I think the ideal use-case I see is that this error pops up in a worker log 3 or more times before an operator has a chance to remediate it, and if the majority of logs blame connector-xyz, the operator can STOP that connector.